src/snyk/common/languageServer/middleware.ts

```diff
@@ -0,0 +1,67 @@
+import {
+  CancellationToken,
+  ConfigurationParams,
+  ConfigurationRequestHandlerSignature,
+  Middleware,
+  ResponseError,
+  WorkspaceMiddleware,
+} from '../vscode/types';
+import { ClientSettings, ServerSettings } from './settings';
+
+export type LanguageClientWorkspaceMiddleware = Partial<WorkspaceMiddleware> & {
+  configuration: (
+    params: ConfigurationParams,
+    token: CancellationToken,
+    next: ConfigurationRequestHandlerSignature,
+  ) => Promise<ResponseError<void> | ServerSettings[]>;
+};
+
+export class LanguageClientMiddleware implements Middleware {
+  workspace: LanguageClientWorkspaceMiddleware = {
+    configuration: async (
+      params: ConfigurationParams,
+      token: CancellationToken,
+      next: ConfigurationRequestHandlerSignature,
+    ) => {
+      let settings = next(params, token);
+      if (this.isThenable(settings)) {
+        settings = await settings;
+      }
+
+      if (settings instanceof Error) {
+        return settings;
+      }
+
+      if (!settings.length) {
+        return [];
+      }
+
+      const clientSettings = settings[0] as ClientSettings;
+      // We should ideally have no setting translation and deliver LS-relevant only.
+      const serverSettings = [
```

Comment:
This looks like it could be merged with `getInitializationOptions` to avoid code duplication.

Reply:
I wish, but they have unfortunately different structure. This is settings.json-based, whereas the other one is our custom configuration singleton that doesn't follow the client settings.json structure.